### ISSUE TITLE
Channels of the optical flow

### ISSUE THREAD

**Issue by vickyliuliu** [AsuradaYuci/SCAN_pytorch#2]
Hello, I ran the codes using the optical flow for ilidsvidsequence dataset you mentioned, but it turned out that the flow  have three channels (RGB). It seems that SCAN requires me to use optical flow images with two channels (x, y). So, I think there is I didn't get right. Can you offer me more information about the optical flow?

**Comment by AsuradaYuci** [comment]
@vickyliuliu Because we save the flows as images (it should has 3 channels), as we all know, the flow information is two channels, So we fixed the third channel with 0.  In the code,When we want to use optical flow images with two channels (x, y), we can just use the specific two channels and discard the useless channel.
In SCAN, we can change the code **reid/models/resnet.py line 73 **  from **motions = motions[:, 0:3]**  to **motions = motions[:, 1:3]  # only choose 2 channels(1,2) for optical flow**.

**Comment by vickyliuliu** [comment]
Thank you for replying me! Just to make sure I understand it right, only the 1, 2 channels are valid x, y, right? 
I tried to change that code in resnet.py to what you have said, but it kept giving me error that when I train my model (specifically it is in the _forward in trainer.py line 106) I need to provide an input with 3 channels. 
Really hope you can help me to figure it out, and thank you for the patience~

**Comment by AsuradaYuci** [comment]
@vickyliuliu when we change the code **reid/models/resnet.py line 73 => motions = motions[:, 1:3]** ,we should also change the code in **reid/models/resnet.py line 36** from **conv0 = nn.Conv2d(3, 64, kernel_size=7, stride=2, padding=3, bias=False)** to **conv0 = nn.Conv2d(2, 64, kernel_size=7, stride=2, padding=3, bias=False)**.

**Comment by vickyliuliu** [comment]
Thanks! I've changed that! 
May I ask another question? When I pass both samplers and labels to the BCE loss function (in /reid/loss/pairloss.py line 49), the loss I get is always NAN. In the terminal I got "reduce failed to synchronize: device-side assert triggered". I searched the internet and I was told that this is because the values of samplers is not in the range of (0,1). But I debugged line by line and found that actually each entry of the samplers tensor is indeed in the (0,1) range, which makes me more confused.

**Comment by AsuradaYuci** [comment]
@vickyliuliu :cry: I have met this problem before, but I forget how to fix it now..

**Comment by vickyliuliu** [comment]
Alright and thank you any way! I will try more to fix it.

**Comment by AsuradaYuci** [comment]
@vickyliuliu good luck :smile:

**Comment by ccq195** [comment]
> Alright and thank you any way! I will try more to fix it.

Hi, did you have solved  this problem? I have met the same problem. @vickyliuliu @AsuradaYuci

**Comment by vickyliuliu** [comment]
@ccq195 Unfortunately I haven't solved it. I'm still trying to figure out. Have you got any idea?

**Comment by ccq195** [comment]
i have add a sigmoid function(samplers = F.sigmoid(samplers)). The error don't appear




------------------ 原始邮件 ------------------
发件人: "vickyliuliu"<notifications@github.com>;
发送时间: 2019年8月7日(星期三) 下午4:42
收件人: "AsuradaYuci/SCAN_pytorch"<SCAN_pytorch@noreply.github.com>;
抄送: "QUIET"<1013757040@qq.com>;"Mention"<mention@noreply.github.com>;
主题: Re: [AsuradaYuci/SCAN_pytorch] Channels of the optical flow (#2)




@ccq195 Unfortunately I haven't solved it. I'm still trying to figure out. Have you got any idea?

**Comment by vickyliuliu** [comment]
I added samplers = sig(samplers), labels = sig(labels) but it still gives the same error...

**Comment by vickyliuliu** [comment]
@ccq195 I tried samplers = sigmoid(samplers) and labels = sigmoid(labels) but the error still appears. This is really strange.

**Comment by ccq195** [comment]
hemm....the labels  don't need sigmoid. you can try only sigmoid the samplers.
------------------ 原始邮件 ------------------
发件人: "vickyliuliu"<notifications@github.com>;
发送时间: 2019年8月7日(星期三) 下午4:49
收件人: "AsuradaYuci/SCAN_pytorch"<SCAN_pytorch@noreply.github.com>;
抄送: "QUIET"<1013757040@qq.com>;"Mention"<mention@noreply.github.com>;
主题: Re: [AsuradaYuci/SCAN_pytorch] Channels of the optical flow (#2)




I added samplers = sig(samplers), labels = sig(labels) but it still gives the same error...

**Comment by vickyliuliu** [comment]
@ccq195  I know...Actually I first tried to only sigmoid the samplers, it didn't work. That's why I also further tried to sigmoid the labels.

**Comment by ccq195** [comment]
so weird! you can print the labels value, analyze it.




------------------ 原始邮件 ------------------
发件人: "vickyliuliu"<notifications@github.com>;
发送时间: 2019年8月7日(星期三) 下午4:59
收件人: "AsuradaYuci/SCAN_pytorch"<SCAN_pytorch@noreply.github.com>;
抄送: "QUIET"<1013757040@qq.com>;"Mention"<mention@noreply.github.com>;
主题: Re: [AsuradaYuci/SCAN_pytorch] Channels of the optical flow (#2)




@ccq195  I know...Actually I first tried to only sigmoid the samplers, it didn't work. That's why I also further tried to sigmoid the labels.

**Comment by vickyliuliu** [comment]
@ccq195 The labels are [0,0,0,0]. I do not know whether this is normal or not?

**Comment by vickyliuliu** [comment]
@ccq195 May I ask what are the values of your labels? I want to see what should be the correct ones

**Comment by ccq195** [comment]
[1,0,0,0,0,1,0,0,....]




------------------ 原始邮件 ------------------
发件人: "vickyliuliu"<notifications@github.com>;
发送时间: 2019年8月7日(星期三) 晚上8:01
收件人: "AsuradaYuci/SCAN_pytorch"<SCAN_pytorch@noreply.github.com>;
抄送: "QUIET"<1013757040@qq.com>;"Mention"<mention@noreply.github.com>;
主题: Re: [AsuradaYuci/SCAN_pytorch] Channels of the optical flow (#2)




@ccq195 May I ask what are the values of your labels? I want to see what should be the correct ones

**Comment by vickyliuliu** [comment]
@ccq195 ok,thank you so much! My labels are likeL [0,0,0,0] (just 4 entries and all 0). I will try to figure out~

**Comment by vickyliuliu** [comment]
@AsuradaYuci I'm stills struggling with the "reduce failed to synchronize" error. I noticed that in .reid/loss/pairloss.py line#23-38, we use a "mask" on the labels. The mask I got is [0,0,0,0], therefore the labels I got is [0,0,0,0] (this is not correct, right? )Could you give a hint on why things are like this?

**Comment by AsuradaYuci** [comment]
@vickyliuliu https://github.com/dapengchen123/video_reid/issues/2  you can see this issue.

**Comment by AsuradaYuci** [comment]
@vickyliuliu If you use my code,I think you will not meet this problem, as I have fixed some problems.

**Comment by vickyliuliu** [comment]
@AsuradaYuci Sorry for disturbing again! Since I'm quite new to computer vision, there are lots of things I don't understand. May I ask what is the use of the "mask"? Why we need to use it?

**Comment by bmiftah** [comment]
Hi, thank you for putting this code here. I I wonder how we can make the model not to use the optical flow and use only image dataset ..

**Comment by AsuradaYuci** [comment]
@bmiftah just check out the inputs,(**inputs = [imgs, flows]**). if we don't want use the optical flow, just change the inputs to (**inputs = [imgs]**),and also we should change somewhere such as (**feat, feat_raw = self.model(inputs[0], inputs[1], self.mode)**) to (**feat, feat_raw = self.model(inputs[0], self.mode)**), and in resnet.py, we don't use **motions** anymore, just comment them.
